File: javascript/rational-numbers/rational-numbers.js
const factors = n => {
  let factors = new Set();
  factors.add(1);
  factors.add(n);
  let i = 2;
  while (i <= Math.sqrt(n)) {
    if (!(n % i)) {
      factors.add(i);
      factors.add(n/i);
    }
    i ++;
  }
  return [...factors];
}

const findGCD = (a, b) => {
  const factorsOfA = factors(a);
  const factorsOfB = factors(b);
  return factorsOfA.reduce((acc, curr) => {
    if (curr > acc && factorsOfB.includes(curr)) return curr;
    return acc;
  }, 1);
}

export class Rational {
  constructor(a, b) {
    // store a and b reduced by greatest common divisor
    if ((a<0 && b <0) || (a>0 && b<0)) {
      a = -a;
      b = -b;
    }
    const gcd = findGCD(Math.abs(a), Math.abs(b));
    
    this.a = a / gcd;
    this.b = this.a ? b / gcd : 1;
  }

  add(n) {
    return new Rational((this.a * n.b) + (this.b * n.a), this.b * n.b);
  }

  sub(n) {
    return new Rational((this.a * n.b) - (this.b * n.a), this.b * n.b);
  }

  mul(n) {
    return new Rational(this.a * n.a, this.b * n.b);
  }

  div(n) {
    return new Rational(this.a * n.b, this.b * n.a);
  }

  abs() {
    return new Rational(this.a >= 0 ? this.a : -1 * this.a, this.b >= 0 ? this.b : -1 * this.b);
  }

  exprational(exp) {
    if (exp >= 0) {
      return new Rational(this.a ** exp, this.b ** exp);
    } else {
      exp = Math.abs(exp);
      return new Rational(this.b ** exp, this.a ** exp);
    }
  }

  expreal(base) {
    // return (base ** (this.a/this.b));
    return Number(Math.pow(base, this.a/this.b).toPrecision(3));
  }

  reduce() {
    return new Rational(this.a, this.b);
  }
}

File: python/rational-numbers/rational_numbers.py
from __future__ import division
from math import gcd, fabs


class Rational:
    def __init__(self, numer, denom):
        divisor = gcd(numer, denom)
        if denom < 0:
            numer = numer * -1
            denom = denom * -1
        self.numer = int(numer / divisor)
        self.denom = int(denom / divisor)

    def __eq__(self, other):
        return self.numer == other.numer and self.denom == other.denom

    def __repr__(self):
        return '{}/{}'.format(self.numer, self.denom)

    def __add__(self, other):
        # (a1 * b2 + a2 * b1) / (b1 * b2)
        numer = self.numer * other.denom + self.denom * other.numer
        denom = self.denom * other.denom
        return Rational(numer, denom)
            

    def __sub__(self, other):
        # (a1 * b2 - a2 * b1) / (b1 * b2)
        numer = self.numer * other.denom - self.denom * other.numer
        denom = self.denom * other.denom
        return Rational(numer, denom)

    def __mul__(self, other):
        # (a1 * a2) / (b1 * b2)
        numer = self.numer * other.numer
        denom = self.denom * other.denom
        return Rational(numer, denom)

    def __truediv__(self, other):
        # (a1 * b2) / (a2 * b1)
        numer = self.numer * other.denom
        denom = self.denom * other.numer
        return Rational(numer, denom)

    def __abs__(self):
        # |a|/|b|
        numer = fabs(self.numer)
        denom = fabs(self.denom)
        divisor = gcd(int(numer), int(denom))
        return Rational(int(numer / divisor), int(denom / divisor))

    def __pow__(self, power):
        if power % 1 == 0:
            if power > 0:
                # (a^n)/(b^n)
                numer = self.numer ** power
                denom = self.denom ** power
            elif power < 0:
                # (b^m)/(a^m), where m = |n|
                numer = self.denom ** fabs(power)
                denom = self.numer ** fabs(power)
            else:
                numer = 1
                denom = 1
        else:
            # (a^x)/(b^x)
            numer = self.numer ** power
            denom = self.denom ** power
        return Rational(numer, denom)

    def __rpow__(self, base):
        # x^(a/b) = root(x^a, b), where root(p, q) is the qth root of p
        if self.numer == 0:
            exponent = 0
        else:
            exponent = float(self.numer / self.denom)
        return base ** exponent
